multiple files: improve gf_dirent_for_name() functionality

mykaul opened the pull request:
- Introduced gf_dirent_for_name2(), which populates most of the fields
with real content and replaced gf_dirent_for_name() in several cases.
- Reduced the no. of calls for strlen() across the code, re-used existing
length fields where applicable.
- Removed dead code

Updates: #3005
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

Commit:
multiple files: improve gf_dirent_for_name() functionality

- Introduced gf_dirent_for_name2(), which populates most of the fields
with real content and replaced gf_dirent_for_name() in several cases.
- Reduced the no. of calls for strlen() across the code, re-used existing
length fields where applicable.
- Removed dead code

Updates: #3005
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

mykaul commented:
/run regression

mykaul commented:
/run regression

mykaul commented:
/run regression

mykaul commented:
/run regression

mykaul commented:
/run regression

amarts reviewed (commented):
LGTM

mykaul commented:
/run regression

mykaul commented:
/run regression

mykaul commented:
Ping?

xhernandez commented on libglusterfs/src/gf-dirent.c:
```diff
@@ -141,26 +141,34 @@ gf_itransform(xlator_t *this, uint64_t x, uint64_t *y_p, int client_id)
 
-    /* TODO: use mem-pool */
-    gf_dirent = GF_CALLOC(gf_dirent_size(name), 1, gf_common_mt_gf_dirent_t);
+    gf_dirent = GF_CALLOC(gf_dirent_len(name_len), 1, gf_common_mt_gf_dirent_t);
```
Couldn't `GF_MALLOC()` be used and just initialize the remaining fields ? this way we avoid writing to `d_name` twice.

mykaul replied:
Sure, I can do that. I just need to memset the iatt struct.

xhernandez commented on libglusterfs/src/glusterfs/gf-dirent.h:
```diff
@@ -15,6 +15,7 @@
 #include "glusterfs/inode.h"
 
 #define gf_dirent_size(name) (sizeof(gf_dirent_t) + strlen(name) + 1)
```
This could be redefined as:
```suggestion
#define gf_dirent_size(name) gf_dirent_len(strlen(name))
```

xhernandez commented on xlators/performance/readdir-ahead/src/readdir-ahead.c:
```diff
@@ -308,22 +302,19 @@ __rda_fill_readdirp(xlator_t *this, gf_dirent_t *entries, size_t request_size,
-
         if (dirent->inode && (!((strcmp(dirent->d_name, ".") == 0) ||
                                 (strcmp(dirent->d_name, "..") == 0)))) {
+            memset(&tmp_stat, 0, sizeof(tmp_stat));
```
I think we can avoid this extra buffer. We could directly use `dirent->d_stat`, and even zero it only if it's not present inside `rda_inode_ctx_get_iatt()`.

mykaul replied:
I think I did exactly what you meant @ https://github.com/gluster/glusterfs/pull/3295 ?

xhernandez replied:
What remains to be done is to completely remove `tmp_stat` variable. You can use `dirent->d_stat` directly.

btw, this PR and #3295 will probably conflict. Which one do you want to merge first ?

mykaul replied:
Honestly, whatever the community feels is ready. This one has been waiting ~3 months, no rush. Whatever is safer, been well reviewed and the community is happy with. I'll resolve conflicts.

xhernandez replied:
I've merged the other PR since it was already reviewed and ready.

xhernandez reviewed (commented):
Just some minor comments...

Commit:
gif-dirent: fix review comments

- improve define
- use MALLOC instead of CALLOC (and initialize relevant fields, so everything is init'ed).

Updates: #3005
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

Commit:
multiple files: improve gf_dirent_for_name() functionality

- Introduced gf_dirent_for_name2(), which populates most of the fields
with real content and replaced gf_dirent_for_name() in several cases.
- Reduced the no. of calls for strlen() across the code, re-used existing
length fields where applicable.
- Removed dead code

Updates: #3005
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/xlators/performance/readdir-ahead/src/readdir-ahead.c b/xlators/performance/readdir-ahead/src/readdir-ahead.c
index 5a666da82..4f3b17c9a 100644
--- a/xlators/performance/readdir-ahead/src/readdir-ahead.c
+++ b/xlators/performance/readdir-ahead/src/readdir-ahead.c
@@ -288,7 +288,7 @@ rda_inode_ctx_get_iatt(inode_t *inode, xlator_t *this, struct iatt *attr)
     }
     UNLOCK(&inode->lock);
 
-    if (ctx_p == NULL) // if we did not find one, return a zero'ed iatt struct.
+    if (ctx_p == NULL)  // if we did not find one, return a zero'ed iatt struct.
         memset(attr, 0, sizeof(struct iatt));
 }
 
```

xhernandez commented:
/run regression

Commit:
readdir-ahead.c: make clang-format happy

Updates: #3005
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

mykaul commented:
/run regression
